### PR TITLE
[packs sdk documentation] upload docs script

### PR DESCRIPTION
**Pull request by spencer-codaio**
foundational script for publishing the docs that are generated by `mkdocs` into the developer documentation S3 bucket.

The file path is explained [here](https://staging.coda.io/d/Scalable-Packs-Storage_dYhioPL2vsk/High-Level-Design_suq7L#_luZc1) under "bucket structure."

image

needs this PR on the infra side for the proper access IAM https://github.com/coda/infra/pull/2125

@coda/packs / @jonathan-codaio @ekoleda-codaio @chris-codaio

### CODE REVIEW

File: Makefile
Comment (jonathan-codaio) [inline Makefile:136]: Nit, do you need the semicolon?

File: Makefile
Comment (jonathan-codaio) [inline Makefile:5]: Any reason not to just do this now?

File: Makefile
Comment (jonathan-codaio) [inline Makefile:141]: The binary is `ts-node`. Also this will only work if node_modules/.bin is on the path, so unless something is setting the path, probably want to reference it as node_modules/.bin/ts-node

File: Makefile
Comment (jonathan-codaio) [inline Makefile:134]: Might want to call this `build-mkdocs`. There's already a `docs` target that "builds" documentation, so we should probably just be specific about what each target actually does.

Alternatively, do we want to add this step to the `docs` target, or is unnecessary to do on each commit?

File: documentation/documentation_publisher.ts
Comment (jonathan-codaio) [inline documentation/documentation_publisher.ts:24]: We have a printAndExit utility that essentially does this, if you want to reuse.

File: documentation/documentation_publisher.ts
Comment (jonathan-codaio) [inline documentation/documentation_publisher.ts:1]: Can use the `print()` utility in this repo that's just a synonym for console.log, that captures the intent of writing to the terminal and avoids needing to deal with the linter.

File: documentation/documentation_publisher.ts
Comment (jonathan-codaio) [inline documentation/documentation_publisher.ts:3]: Nit, I know `coda` just eliminated yargs in favor of Commander, but we only use yargs in this repo right now. Would be nice to migrate the the whole repo to Commander, but we have a lot bigger fish to fry right now, so would suggest just using yargs so we don't have to deal with two libraries for commands, and we can migrate the whole repo all at once at some later date.

File: documentation/documentation_publisher.ts
Comment (jonathan-codaio) [inline documentation/documentation_publisher.ts:64]: No idea if it makes sense here, but in the `packs` repo we use the `s3 sync` command, which can handle syncing an entire directory full of content to/from S3 all in one go. We use this in the CI to pull down existing packs and do diffs against the latest build.

https://github.com/coda/packs/blob/main/scripts/get-current-packs.sh#L22

File: documentation/documentation_publisher.ts
Comment (alan-codaio) [inline documentation/documentation_publisher.ts:88]: I believe we also have a `print` helper, but if your code isn't triggering our no-console lint rule then it's probably fine to leave as is?

File: documentation/documentation_publisher.ts
Comment (jonathan-codaio) [reply documentation/documentation_publisher.ts:3]: Yeah, these things tend to sit in an intermediate state for months. It's not great if you're trying to make simple changes to logic and you find yourself having to learn a whole new library just because the repo is inconsistent.

Comment (spencer-codaio) [reply documentation/documentation_publisher.ts:3]: ok done

**Comment by spencer-codaio** [comment]
ok updated to using aws s3 sync and the standard yargs stuff

image

will follow-up with adding the makefile handling for `activate` command

File: documentation/documentation_publisher.ts
Comment (jonathan-codaio) [inline documentation/documentation_publisher.ts:49]: `spawnSync` may work here too, not sure if there's a need to make async vs not

File: documentation/documentation_publisher.ts
Comment (jonathan-codaio) [inline documentation/documentation_publisher.ts:67]: Probably want these to get pushed in a deterministic order, and let one succeed before trying the others? In which case, could use spawnSync and just use a regular `for` loop.

`prod` needs to get split out into a separate step I think though, in order for CircleCI to halt and wait for approval first?

File: documentation/documentation_publisher.ts
Comment (jonathan-codaio) [inline documentation/documentation_publisher.ts:25]: Have you been able to test this locally? If so, is there a way to print a failure message if the credentials are invalid/missing, that explains what a developer should do to set up their environment to test this?

Comment (spencer-codaio) [reply documentation/documentation_publisher.ts:25]: yeah i've been testing this locally and sure thing! I believe `prodaccess-admin` inserts the appropriate credentials into the ~/.aws/credentials file so i can try once my credentials are expired to see what the error is

File: documentation/documentation_publisher.ts
Comment (spencer-codaio) [reply documentation/documentation_publisher.ts:49]: oh interesting didnt know a sync version existed will try that

File: documentation/documentation_publisher.ts
Comment (spencer-codaio) [reply documentation/documentation_publisher.ts:67]: sure that sounds reasonable for pushing them in order.

the reason they all get pushed at the same time (including prod) is because the runtime config flow is 1) push everything 2) progressively active the new version for each environment with activating the new prod waiting on approval. Will see how feasible we can use the "active" piece with Cloudfront serving and then revisit the flow here if not.

Comment (spencer-codaio) [reply documentation/documentation_publisher.ts:67]: although I guess if we're making it possible to serve all versions at once we actually want the push step to be hidden too since it might accidentally expose the latest version that hasn't been approved if someone is able to enumerate the URL so maybe just cleaner to only push once we approve too

Comment (spencer-codaio) [reply documentation/documentation_publisher.ts]: ok this is resolved with the new logic

**Comment by spencer-codaio** [comment]
New flow removes the "activating" logic and just relies on "pushing" as the way to release. Then in the CI workflow, we'll "push" progressively from head -> staging -> wait for approval -> prod.

image

File: documentation/documentation_publisher.ts
Comment (jonathan-codaio) [inline documentation/documentation_publisher.ts:18]: Total nice, `${e.message} ${e.stack || ''}`

File: documentation/documentation_publisher.ts
Comment (jonathan-codaio) [inline documentation/documentation_publisher.ts:10]: Nit, newline above.

File: package.json
Comment (jonathan-codaio) [inline package.json:18]: Can this go in devDependencies with all the other types?

File: documentation/documentation_publisher.ts
Comment (jonathan-codaio) [inline documentation/documentation_publisher.ts:60]: I'm not sure if this is that useful if we don't stop you if you're going to overwrite. I think it's fine to omit this, but if we want to warn people then we should error out in this case unless you pass a commandline flag to force the upload.

**Review by jonathan-codaio — COMMENTED** [review]
Nice, quite clean now. Just minor stuff.

File: documentation/documentation_publisher.ts
Comment (jonathan-codaio) [reply documentation/documentation_publisher.ts:65]: That's fine, mind having it error out here then, and have a yargs flag to force it to continue?

**Review by jonathan-codaio — APPROVED** [review]
How does one use `--forceUpload` in practice though, do you just add it to `make publish-docs-head --forcePublish` etc?

What happens if the CI stalls because of this? Assuming the plan is that the Codan just runs it manually on a local machine to overwrite and the CI is never supposed to force upload. Would just be good to comment about the intended usage, maybe near the flag definition or something.

**Pull request merged**
